schema_registry.cc

```diff
@@ -274,39 +275,73 @@ global_schema_ptr::global_schema_ptr(global_schema_ptr&& o) noexcept {
-        // because _ptr is preventing this.
-        const schema_registry_entry& e = *_ptr->registry_entry();
-        schema_ptr s = local_schema_registry().get_or_null(e.version());
-        if (!s) {
-            s = local_schema_registry().get_or_load(e.version(), [&e](table_schema_version) {
-                return e.frozen();
-            });
+        auto registered_schema = [](const schema_registry_entry& e) {
+            schema_ptr ret = local_schema_registry().get_or_null(e.version());
+            if (!ret) {
+                ret = local_schema_registry().get_or_load(e.version(), [&e](table_schema_version) {
+                    return e.frozen();
+                });
+            }
+            return ret;
+        };
+
+        schema_ptr registered_bs;
+        // the following code contains registry entry dereference of a foreign shard
+        // however, it is guarantied to succeed since we made sure in the constructor
+        // that _bs_schema and _ptr will have a registry on the foreign shard where this
+        // object originated so as long as this object lives the registry entries lives too
+        // and it is safe to reference them on foreign shards.
+        if (_base_schema) {
+            registered_bs = registered_schema(*_base_schema->registry_entry());
+            if (_base_schema->registry_entry()->is_synced()) {
+                registered_bs->registry_entry()->mark_synced();
+            }
         }
-        if (e.is_synced()) {
+        schema_ptr s = registered_schema(*_ptr->registry_entry());
+        if (s->is_view()) {
+            if (! s->view_info()->base_info()) {
+                // we know that registered_bs is valid here because we make sure of it in the constructors.
+                s->view_info()->set_base_info(s->view_info()->make_base_dependent_view_info(*registered_bs));
+            }
+        }
+        if (_ptr->registry_entry()->is_synced()) {
             s->registry_entry()->mark_synced();
         }
         return s;
     }
 }
 
 global_schema_ptr::global_schema_ptr(const schema_ptr& ptr)
-    : _ptr([&ptr]() {
-        // _ptr must always have an associated registry entry,
-        // if ptr doesn't, we need to load it into the registry.
-        schema_registry_entry* e = ptr->registry_entry();
-        if (e) {
-            return ptr;
+        : _cpu_of_origin(this_shard_id()) {
+    // _ptr must always have an associated registry entry,
+    // if ptr doesn't, we need to load it into the registry.
+    schema_registry_entry* e = ptr->registry_entry();
+    schema_ptr s;
+    if (e) {
+        s = ptr;
+    } else {
+        s = local_schema_registry().get_or_load(ptr->version(), [&ptr] (table_schema_version) {
+            return frozen_schema(ptr);
+        });
+    }
+    if (s->is_view()) {
+        if (s->view_info()->base_info()) {
+            _base_schema = global_schema_ptr(s->view_info()->base_info()->base_schema());
```

Comment:
The type of _base_schema is schema_ptr, why do you assign it via global_schema_ptr ?

Reply:
This is kind of code deduplication, it will run the base schema through this constructor. One of this constructors jobs
is to make sure that the schema pointer has a registry entry, this is what makes it permanent. So instead of using the
same code from lines 322-332 I just ran it through the constructor. But now that you mention it, I will deduplicate it
differently using a lambda.